lib/3scale/backend/storage_async/client.rb

```diff
@@ -0,0 +1,188 @@
+      class Client
+        include Configurable
+
+        DEFAULT_HOST = 'localhost'.freeze
+        private_constant :DEFAULT_HOST
+
+        DEFAULT_PORT = 22121
+        private_constant :DEFAULT_PORT
+
+        HOST_PORT_REGEX = /redis:\/\/(.*):(\d+)/
+        private_constant :HOST_PORT_REGEX
+
+        class << self
+          def instance(reset = false)
+            if reset || @instance.nil?
+              @instance = new(configuration.redis)
+            else
+              @instance
+            end
+          end
+        end
+
+        def initialize(opts)
+          host, port = opts[:url].match(HOST_PORT_REGEX).captures if opts[:url]
+          host ||= DEFAULT_HOST
+          port ||= DEFAULT_PORT
+
+          endpoint = Async::IO::Endpoint.tcp(host, port)
+          @redis_async = Async::Redis::Client.new(endpoint)
+        end
+
+        # Now we are going to define the methods to run redis commands
+        # following the interface of the redis-rb lib.
+        #
+        # These are the different cases:
+        #   1) Methods that can be called directly. For example SET:
+        #      @redis_async.call('SET', some_key)
+        #   2) Methods that need to be "boolified". These are methods for which
+        #      redis-rb returns a boolean, but redis just returns an integer.
+        #      For example, Redis returns 0 or 1 for the EXISTS command, but
+        #      redis-rb transforms that into a boolean.
+        #   3) Methods that accept an array as their first parameter in
+        #      redis-rb. For example, the DEL command accepts an array of keys.
+        #   4) Methods that accept an array as their second parameter. For
+        #      example, LPUSH accepts an array of keys.
+        #   5) There are a few methods that need to be treated differently and
+        #      do not fit in any of the previous categories. For example, SSCAN
+        #      which accepts a hash of options in redis-rb.
+        #
+        # All of this might be simplified a bit in the future using the
+        # "methods" in async-redis
+        # https://github.com/socketry/async-redis/tree/master/lib/async/redis/methods
+        # but there are some commands missing, so for now, that's not an option.
+
+        METHODS_TO_BE_CALLED_DIRECTLY = [
+          :expire,
+          :expireat,
+          :flushdb,
+          :get,
+          :hset,
+          :hmget,
+          :incr,
+          :incrby,
+          :keys,
+          :llen,
+          :lpop,
+          :lrange,
+          :ltrim,
+          :scard,
+          :setex,
+          :smembers,
+          :ttl,
+          :zcard,
+          :zrangebyscore,
+          :zremrangebyscore,
+          :zrevrange
+        ].freeze
+        private_constant :METHODS_TO_BE_CALLED_DIRECTLY
+
+        METHODS_TO_BE_CALLED_DIRECTLY.each do |method|
+          define_method(method) do |*args|
+            @redis_async.call(method.to_s.upcase, *args)
+          end
+        end
```

Comment:
`async-redis` offers 2 different interfaces for making requests:
1) Perform redis calls with .call(): `client.call('GET', 'k')`, `client.call('SET', 'k', '1')`, etc. This works for any redis operation.
2) Perform redis calls with methods like `client.get('k')`, `client.set('k', '1')`, etc. In the library they refer to these methods as "dsl" or "methods". They are defined here: https://github.com/socketry/async-redis/tree/master/lib/async/redis/methods The problem is that not all the commands are supported (for example, they do not provide any methods for sets). Also, those methods don't necessarily need to provide the same interface as `redis-rb`, so we might end up doing something similar to what we have now anyway. I'll leave this as it is now. In the future, I think it might make sense to contribute to the library to implement the methods that are missing.

Reply:
We are still experimenting with how to best support all Redis methods, we experimenting with generating method stubs from redis documentation.